File: packages/go/subaccounts.go
package hubfly

import (
	"context"
	"fmt"
)

type SubaccountsService struct{ client *Client }

func (s *SubaccountsService) List(ctx context.Context) (*Response[[]Subaccount], error) {
	return Request[[]Subaccount](ctx, s.client, "GET", "/api/v1/subaccounts", nil)
}

func (s *SubaccountsService) Create(ctx context.Context, params CreateSubaccountParams) (*Response[Subaccount], error) {
	return Request[Subaccount](ctx, s.client, "POST", "/api/v1/subaccounts/create", params)
}

func (s *SubaccountsService) Get(ctx context.Context, id string) (*Response[Subaccount], error) {
	return Request[Subaccount](ctx, s.client, "GET", fmt.Sprintf("/api/v1/subaccounts/%s", id), nil)
}

func (s *SubaccountsService) Suspend(ctx context.Context, id string) (*Response[Subaccount], error) {
	return Request[Subaccount](ctx, s.client, "POST", fmt.Sprintf("/api/v1/subaccounts/%s/suspend", id), nil)
}

func (s *SubaccountsService) Resume(ctx context.Context, id string) (*Response[Subaccount], error) {
	return Request[Subaccount](ctx, s.client, "POST", fmt.Sprintf("/api/v1/subaccounts/%s/resume", id), nil)
}

func (s *SubaccountsService) Close(ctx context.Context, id string) (*Response[Subaccount], error) {
	return Request[Subaccount](ctx, s.client, "POST", fmt.Sprintf("/api/v1/subaccounts/%s/close", id), nil)
}

func (s *SubaccountsService) ListKeys(ctx context.Context, id string) (*Response[[]SubaccountKey], error) {
	return Request[[]SubaccountKey](ctx, s.client, "GET", fmt.Sprintf("/api/v1/subaccounts/%s/api-keys", id), nil)
}

func (s *SubaccountsService) CreateKey(ctx context.Context, id string, params map[string]interface{}) (*Response[SubaccountKey], error) {
	return Request[SubaccountKey](ctx, s.client, "POST", fmt.Sprintf("/api/v1/subaccounts/%s/api-keys/create", id), params)
}

func (s *SubaccountsService) RevokeKey(ctx context.Context, id, keyID string) (*Response[map[string]bool], error) {
	return Request[map[string]bool](ctx, s.client, "POST", fmt.Sprintf("/api/v1/subaccounts/%s/api-keys/%s/revoke", id, keyID), nil)
}

File: packages/ts/src/modules/subaccounts.ts
import type { HubflyClient } from '../client.js';
import type { ApiResponse, Subaccount, SubaccountKey } from '../types.js';

export class SubaccountsModule {
  constructor(private readonly client: HubflyClient) {}

  public async list(): Promise<ApiResponse<Subaccount[]>> {
    return this.client.request<Subaccount[]>('/api/v1/subaccounts');
  }

  public async create(params: { name: string; externalRef?: string; metadata?: Record<string, string>; parentType?: 'user' | 'organization'; parentId?: string }): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>('/api/v1/subaccounts/create', { method: 'POST', body: params });
  }

  public async get(subaccountId: string): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>(`/api/v1/subaccounts/${subaccountId}`);
  }

  public async update(subaccountId: string, params: { name?: string; metadata?: Record<string, string>; maxProjects?: number | null; spendLimitMicroUsd?: string | null }): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>(`/api/v1/subaccounts/${subaccountId}`, { method: 'POST', body: params });
  }

  public async suspend(subaccountId: string): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>(`/api/v1/subaccounts/${subaccountId}/suspend`, { method: 'POST' });
  }

  public async resume(subaccountId: string): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>(`/api/v1/subaccounts/${subaccountId}/resume`, { method: 'POST' });
  }

  public async close(subaccountId: string): Promise<ApiResponse<Subaccount>> {
    return this.client.request<Subaccount>(`/api/v1/subaccounts/${subaccountId}/close`, { method: 'POST' });
  }

  public async listKeys(subaccountId: string): Promise<ApiResponse<SubaccountKey[]>> {
    return this.client.request<SubaccountKey[]>(`/api/v1/subaccounts/${subaccountId}/api-keys`);
  }

  public async createKey(subaccountId: string, params: { name?: string; scopes?: string[]; expiresAt?: string | null } = {}): Promise<ApiResponse<SubaccountKey>> {
    return this.client.request<SubaccountKey>(`/api/v1/subaccounts/${subaccountId}/api-keys/create`, { method: 'POST', body: params });
  }

  public async revokeKey(subaccountId: string, keyId: string): Promise<ApiResponse<{ ok: boolean }>> {
    return this.client.request<{ ok: boolean }>(`/api/v1/subaccounts/${subaccountId}/api-keys/${keyId}/revoke`, { method: 'POST' });
  }

  public async wallet(subaccountId: string): Promise<ApiResponse<{ subaccountId: string; balanceMicroUsd: string; status: string }>> {
    return this.client.request(`/api/v1/subaccounts/${subaccountId}/wallet`);
  }

  public async transfer(subaccountId: string, amountMicroUsd: string, idempotencyKey: string): Promise<ApiResponse<{ id: string; subaccountId: string; amountMicroUsd: string; status: string }>> {
    return this.client.request(`/api/v1/subaccounts/${subaccountId}/wallet/transfers`, { method: 'POST', body: { amountMicroUsd, idempotencyKey } });
  }
}
